packages/component-base/src/data-provider-controller/data-provider-controller.d.ts

```diff
@@ -0,0 +1,148 @@
+/**
+ * @license
+ * Copyright (c) 2021 - 2023 Vaadin Ltd.
+ * This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
+ */
+import type { ReactiveController } from 'lit';
+import type { Cache } from './cache.js';
+
+type DataProviderParams = {
+  page: number;
+  pageSize: number;
+  parentItem?: unknown;
+  [key: string]: unknown;
+};
+
+type DataProviderCallback = (items: unknown[], size?: number) => void;
+
+type DataProvider = (params: DataProviderParams, callback: DataProviderCallback) => void;
+
+/**
+ * A controller that stores and manages items loaded with a data provider.
+ */
+export class DataProviderController implements ReactiveController {
+  /**
+   * The controller host element.
+   */
+  host: HTMLElement;
+
+  /**
+   * A callback that returns data based on the passed params such as
+   * `page`, `pageSize`, `parentItem`, etc.
+   */
+  dataProvider: DataProvider;
+
+  /**
+   * A callback that returns additional params that need to be passed
+   * to the data provider callback with every request.
+   */
+  dataProviderParams: () => Record<string, unknown>;
+
+  /**
+   * A number of items in the root cache.
+   */
+  size?: number;
+
+  /**
+   * A number of items to display per page.
+   */
+  pageSize: number;
+
+  /**
+   * A callback that returns whether the given item is expanded.
+   */
+  isExpanded: (item: unknown) => boolean;
+
+  /**
+   * A reference to the root cache instance.
+   */
+  rootCache: Cache;
+
+  constructor(
+    host: DataProviderController['host'],
+    params: {
+      size: DataProviderController['size'];
+      pageSize: DataProviderController['pageSize'];
+      isExpanded: DataProviderController['isExpanded'];
+      dataProvider: DataProviderController['dataProvider'];
+      dataProviderParams: DataProviderController['dataProviderParams'];
+    },
+  ): void;
```

Comment:
Getting this error:
<img width="542" alt="Screenshot 2023-09-25 at 11 04 43" src="https://github.com/vaadin/web-components/assets/1222264/98df1726-5a35-46b8-bf3d-41f6d3af8992">

Reply:
Fixed.